python/ml4ir/data/tfrecord_reader.py

```
@@ -79,72 +88,102 @@ def _parse_sequence_example_fn(sequence_example_proto):
+            num_records = tf.cast(tf.reduce_sum(mask), tf.int64)
+
+            if pad_records:
+                mask = tf.expand_dims(mask, axis=-1)
+
+                def crop_fn():
+                    tf.print("\n[WARN] Bad query found. Number of records : ", tf.shape(mask)[1])
+                    return image.crop_to_bounding_box(
+                        mask,
+                        offset_height=0,
+                        offset_width=0,
+                        target_height=1,
+                        target_width=max_num_records,
+                    )
+
+                mask = tf.cond(
+                    tf.shape(mask)[1] < max_num_records,
+                    # Pad if there are missing records
+                    lambda: image.pad_to_bounding_box(
+                        mask,
+                        offset_height=0,
+                        offset_width=0,
+                        target_height=1,
+                        target_width=max_num_records,
+                    ),
+                    # Crop if there are extra records
+                    crop_fn,
+                )
+                mask = tf.squeeze(mask)
+            else:
+                mask = tf.squeeze(mask, axis=0)
+
+            # Check validity of mask
+            tf.debugging.assert_greater(num_records, tf.constant(0, dtype=tf.int64))
+
+            features_dict["mask"] = mask
+            num_records = max_num_records if pad_records else num_records
+
         # Pad sequence features to max_num_records
         for feature_info in feature_config.get_sequence_features():
             feature_node_name = feature_info.get("node_name", feature_info["name"])
             feature_layer_info = feature_info["feature_layer_info"]
             preprocessing_info = feature_info.get("preprocessing_info", {})
 
-            feature_tensor = sequence_features.get(feature_node_name)
+            default_tensor = tf.fill(
```

@jakemannix For fields that were not passed in the proto, we fill it with their default value. This would be fields that not `required`.

---

Were we wanting to do this? I’ve been interpreting  ‘required’ as meaning the SequenceExample has to have the feature.  You should possibly be doing an assert() that the field is already filled in, not fill it in yourself, right?

---

This is for fields that the model needs, say something like `organization_id`, but which is not a required field and will NOT be in the input proto at inference.

I should do an assert though. That might be a good way to check we are not silently failing.

---

Going back to this
- There is no need for an explicit assert. If the required feature is not present, then `tf.io.parse_single_sequence_example()` will blow up.
- Might be difficult/impossible to do an assert to check presence of a dictionary key using `tf.debugging`